### PR TITLE
Add screenshot utility

### PR DESCRIPTION
**Pull request by BahdanSerabrakou**
#50

### CODE REVIEW

**Review by ValiantsinSerada — DISMISSED** [review]

File: src/Mafin.Web.UI.Selenium/Mafin.Web.UI.Selenium.csproj [hunk 1]
```diff
@@ -15,7 +15,11 @@
 
     <PackageReference Include="Selenium.WebDriver" />
     <PackageReference Include="Selenium.Support" />
+    <PackageReference Include="NUnit" />
```
Comment (Molnix888) [inline src/Mafin.Web.UI.Selenium/Mafin.Web.UI.Selenium.csproj:18]: Let's provide uour users with the flexibility of choosing test runner on their own. We shouldn't have a hard dependency on NUnit.

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj [hunk 2]
```diff
@@ -2,17 +2,23 @@
   <PropertyGroup>
     <TargetFramework>net7.0</TargetFramework>
     <IsPackable>false</IsPackable>
+    <Nullable>warnings</Nullable>
```
Comment (Molnix888) [inline tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj:5]: Could you please elaborate why need for the flag?

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj [hunk 3]
```diff
@@ -2,17 +2,23 @@
...
+    <PackageReference Include="Selenium.Support" />
+    <PackageReference Include="Selenium.WebDriver.Extensions" />
     <PackageReference Include="xunit" />
     <PackageReference Include="xunit.runner.visualstudio" PrivateAssets="all" IncludeAssets="runtime; build; native; contentfiles; analyzers" />
     <PackageReference Include="coverlet.collector" PrivateAssets="all" IncludeAssets="runtime; build; native; contentfiles; analyzers" />
 
     <ProjectReference Include="..\..\src\Mafin.Web.UI.Selenium\Mafin.Web.UI.Selenium.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="TestResults\" />
```
Comment (Molnix888) [inline tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj:22]: Do we need this folder explicitly?

**Review by Molnix888 — CHANGES_REQUESTED** [review]

File: Directory.Packages.props [hunk 4]
```diff
@@ -11,6 +11,7 @@
     <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.7.0" />
     <PackageVersion Include="Moq" Version="4.18.4" />
     <PackageVersion Include="Philips.CodeAnalysis.DuplicateCodeAnalyzer" Version="1.1.7" />
+    <PackageVersion Include="Selenium.WebDriver.Extensions" Version="4.1.0" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:14]: Fixed

File: Directory.Packages.props [hunk 5]
```diff
@@ -21,5 +22,7 @@
     <PackageVersion Include="Selenium.Support" Version="4.11.0" />
     <PackageVersion Include="WebDriverManager" Version="2.16.3" />
     <PackageVersion Include="Yapoml.Selenium" Version="0.36.4" />
+    <PackageVersion Include="Noksa.WebDriver.ScreenshotsExtensions" Version="0.1.5.4" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:25]: Fixed

File: Directory.Packages.props (hunk 4 above)
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:14]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
+using System.Drawing;
+using Mafin.Web.UI.Selenium.Enums;
+using NUnit.Framework;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.Extensions;
+using WDSE;
+using WDSE.Decorators;
+using WDSE.ScreenshotMaker;
+
+namespace Mafin.Web.UI.Selenium.Utilities
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs:10]: Fixed

File: src/Mafin.Web.UI.Selenium/Mafin.Web.UI.Selenium.csproj (hunk 1 above)
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Mafin.Web.UI.Selenium.csproj:18]: Fixed

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj (hunk 3 above)
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj:22]: Fixed

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj (hunk 2 above)
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj:5]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        /// <param name="element">Element to take.</param>
+        /// <param name="screenshotName">Name of a screenshot file.</param>
+        /// <param name="screenshotDirectory">Directory to save a screenshot file.</param>
+        /// <param name="isAttached">If true - screenshot is atteched to the test result, false otherwise.</param>
+        /// <param name="elementsToHide">Elements to hide on a fullscreen picture.</param>
+        public static void TakeFlexibleScreenshot(this IWebDriver driver, ScreenshotType screenshotType = ScreenshotType.DefaultScreen, IWebElement? element = null, string screenshotName = "", string screenshotDirectory = "Screenshots", bool isAttached = false, params By[] elementsToHide)
+        {
+            if (!Directory.Exists(screenshotDirectory))
+            {
+                Directory.CreateDirectory(screenshotDirectory);
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs:31]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+using WDSE;
+using WDSE.Decorators;
+using WDSE.ScreenshotMaker;
+
+namespace Mafin.Web.UI.Selenium.Utilities
+{
+    /// <summary>
+    /// Give possibility to take customizible screenshot.
+    /// </summary>
+    public static class ScreenshotUtility
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/ScreenshotUtility.cs:15]: Fixed

File: Directory.Packages.props [hunk 9]
```diff
@@ -21,5 +22,7 @@
...
+    <PackageVersion Include="System.Drawing.Common" Version="7.0.0" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:26]: Fixed

File: Directory.Packages.props (hunk 5 above)
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
+using ImageMagick;
```
Comment (Molnix888) [inline src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs:1]: I'm wondering if both fullscreenshot and element screenshot actually possible to be created using only single additional dependency: either ImageMagick or SkiaSharp. On the first sight seems both should be capable of doing so, so I'm thinking whether we should try.

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
+using Mafin.Web.UI.Selenium.Enums;
+using Mafin.Web.UI.Selenium.Utilities;
+using Moq;
+using OpenQA.Selenium;
+
+namespace Mafin.Web.UI.Selenium.Tests.Unit.Utilities
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs:6]: Fixed

File: Directory.Packages.props [hunk 12]
```diff
@@ -7,12 +7,14 @@
...
     <PackageVersion Include="FluentAssertions" Version="6.12.0" />
     <PackageVersion Include="FluentAssertions.Analyzers" Version="0.24.0" />
+    <PackageVersion Include="Magick.NET-Q16-AnyCPU" Version="13.2.0" />
     <PackageVersion Include="Microsoft.CST.DevSkim" Version="1.0.20" />
     <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.7.2" />
     <PackageVersion Include="Moq" Version="4.18.4" />
     <PackageVersion Include="nunit" Version="3.13.3" />
     <PackageVersion Include="NUnit3TestAdapter" Version="4.5.0" />
     <PackageVersion Include="Philips.CodeAnalysis.DuplicateCodeAnalyzer" Version="1.1.7" />
+    <PackageVersion Include="SkiaSharp" Version="2.88.5" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:17]: Thanks!

File: Directory.Packages.props [hunk 13]
```diff
@@ -7,12 +7,14 @@
     <PackageVersion Include="coverlet.collector" Version="6.0.0" />
     <PackageVersion Include="FluentAssertions" Version="6.12.0" />
     <PackageVersion Include="FluentAssertions.Analyzers" Version="0.24.0" />
+    <PackageVersion Include="Magick.NET-Q16-AnyCPU" Version="13.2.0" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply Directory.Packages.props:10]: Thanks!

File: examples/Mafin.Web.UI.Selenium.Example/Mafin.Web.UI.Selenium.Example.csproj [hunk 14]
```diff
@@ -7,6 +7,7 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="NUnit" />
     <PackageReference Include="NUnit3TestAdapter" />
+    <PackageReference Include="SkiaSharp" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply examples/Mafin.Web.UI.Selenium.Example/Mafin.Web.UI.Selenium.Example.csproj:10]: Fixed

File: examples/Mafin.Web.UI.Selenium.YapomlExample/Mafin.Web.UI.Selenium.YapomlExample.csproj [hunk 15]
```diff
@@ -7,6 +7,8 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="NUnit" />
     <PackageReference Include="NUnit3TestAdapter" />
+    <PackageReference Include="Selenium.Support" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply examples/Mafin.Web.UI.Selenium.YapomlExample/Mafin.Web.UI.Selenium.YapomlExample.csproj:10]: Fixed

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj [hunk 16]
```diff
@@ -10,9 +10,11 @@
     <PackageReference Include="FluentAssertions.Analyzers" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="Moq" />
+    <PackageReference Include="SkiaSharp" />
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Mafin.Web.UI.Selenium.Tests.Unit.csproj:13]: Fixed

File: src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs [hunk 17]
```diff
@@ -0,0 +1,8 @@
+namespace Mafin.Web.UI.Selenium.Enums;
+
+public enum ScreenshotType
+{
+    FullScreen,
+    SingleElement,
+    DefaultScreen
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs:7]: Fixed

File: src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs [hunk 18]
```diff
@@ -0,0 +1,8 @@
+namespace Mafin.Web.UI.Selenium.Enums;
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs:1]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+using Mafin.Web.UI.Selenium.Enums;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.Extensions;
+
+namespace Mafin.Web.UI.Selenium.Utilities;
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs:6]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+/// <summary>
+/// Presents screenshot extension methods for <see cref="IWebDriver"/>.
+/// </summary>
+public static class DriverScreenshotExtensions
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs:11]: Fixed

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    private const string ScreenshotsFileName = "testscreenshot";
+
+    public void Dispose() => Directory.Delete(ScreenshotsFolderName, true);
+
+    [Theory]
+    [InlineData("png")]
+    [InlineData("jpg")]
+    public void TakeFlexibleScreenshot_WhenTakeScreensotWithDifferentExtensions_ShouldSaveWithCorrespondingExtension(string extension)
+    {
+        // Arrange
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs:20]: Fixed

File: tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+        // Act
+        driverMock.Object.TakeFlexibleScreenshot(ScreenshotType.DefaultScreen, elementMock.Object, $"{ScreenshotsFileName}.{extension}", ScreenshotsFolderName, elementsToHide.ToArray());
+
+        // Assert
+        File.Exists(elementScreenshotPath).Should().BeTrue();
+    }
+
+    [Fact]
+    public void TakeFlexibleScreenshot_WhenTakeDefaultScreenshot_ShouldSaveScreenshot()
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply tests/Mafin.Web.UI.Selenium.Tests.Unit/Utilities/DriverScreenshotExtensionsTests.cs:39]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+            image.Write(filePath);
+        }
+    }
+
+    private static void TakeFullPageScreenshot(this IWebDriver driver, string screenshotFilePath = "", params By[] elementsLocatorToHide)
+    {
+        long windowHeight = (long)((IJavaScriptExecutor)driver).ExecuteScript("return window.innerHeight;");
+        long totalHeight = (long)((IJavaScriptExecutor)driver).ExecuteScript("return document.documentElement.scrollHeight;");
+
+        var screenshotCollection = new MagickImageCollection();
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+        using (var image = new MagickImage(screenshotStream))
+        {
+            var cropRectangle = new MagickGeometry(element.Location.X, element.Location.Y, element.Size.Width, element.Size.Height);
+
+            image.Crop(cropRectangle);
+            image.Write(filePath);
+        }
+    }
+
+    private static void TakeFullPageScreenshot(this IWebDriver driver, string screenshotFilePath = "", params By[] elementsLocatorToHide)
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs:53]: Fixed

File: src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+
+public enum ScreenshotType
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Enums/ScreenshotType.cs:3]: Fixed

File: src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+    /// <summary>
+    /// Customizible screenshot method.
+    /// </summary>
+    /// <param name="driver">Webdriver instance.</param>
+    /// <param name="screenshotType">Type of a screenshot.</param>
+    /// <param name="element">Element to take.</param>
+    /// <param name="screenshotName">Name of a screenshot file.</param>
+    /// <param name="screenshotDirectory">Directory to save a screenshot file.</param>
+    /// <param name="elementsLocatorToHide">Elements to hide on a fullscreen picture.</param>
+    public static void TakeFlexibleScreenshot(this IWebDriver driver, ScreenshotType screenshotType = ScreenshotType.DefaultScreen, IWebElement? element = null, string screenshotName = "", string screenshotDirectory = "Screenshots", params By[] elementsLocatorToHide)
```
[the remark this answers is not in the document]
Comment (BahdanSerabrakou) [reply src/Mafin.Web.UI.Selenium/Utilities/DriverScreenshotExtensions.cs:22]: Fixed
Fixed

**Review by Molnix888 — APPROVED** [review]

**Pull request merged**